Remove ProcessSyncDataType::HasFullscreenElement

achristensen07 opened the pull request:
#### ae35ee5cdb6f727e4e8f6ddffc1aeebb2cbb769f
<pre>
Remove ProcessSyncDataType::HasFullscreenElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=287541">https://bugs.webkit.org/show_bug.cgi?id=287541</a>
<a href="https://rdar.apple.com/144667685">rdar://144667685</a>

Reviewed by Tim Nguyen.

We don&apos;t need all site isolation processes to know that the top document has
a fullscreen element.  We just need to check if the fullscreen element in this
process is in the top document when a request comes to exit fullscreen.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::populateDocumentSyncDataForNewlyConstructedDocument):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::clearFullscreenFlags):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::updatePageFullscreenStatusIfTopDocument): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateProcessSyncData):
(WebCore::Page::setTopDocumentHasFullscreenElement): Deleted.
(WebCore::Page::topDocumentHasFullscreenElement): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::requestExitFullScreen):

Canonical link: <a href="https://commits.webkit.org/290288@main">https://commits.webkit.org/290288@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed341195f57fc759a5b628b031361b9f67415de2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89570 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9097 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44429 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94562 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40337 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91622 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9486 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17376 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/68998 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/26644 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92571 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/7277 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/49362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/7008 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39443 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/77362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/36675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96390 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16752 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77872 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17007 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/77089 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/77188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/126/builds/21611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20191 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9919 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/14048 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16765 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/16506 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19957 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/18288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
As much as I'd love getting rid of this code, fullscreen/iframe-fullscreen-system-exit.html seems to be regressing (e.g. https://bugs.webkit.org/show_bug.cgi?id=287309 )

nt1m commented on Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
```diff
@@ -547,7 +547,7 @@ void WebFullScreenManager::requestExitFullScreen()
     }
 
-    if (!corePage->topDocumentHasFullscreenElement()) {
+    if (!m_element || !m_element->document().isTopDocument()) {
```
I'm guessing there's a mismatch between `WebFullscreenManager::m_element` and `FullscreenManager::fullscreenElement()`. If I had to guess, it's the same mismatch than between `FullscreenManager::fullscreenElement()` and `FullscreenManager::m_fullscreenElement`

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -8582,6 +8581,16 @@ CheckedRef<const FullscreenManager> Document::checkedFullscreenManager() const
+    }
+
+    return nullptr;
+}
```
Any reason to move this to Document ? Can't we just leave this on FullscreenManager and access it from there? I'd prefer fullscreen logic to be contained.

nt1m reviewed (approved):
LGTM modulo leaving the move to Document.cpp

achristensen07 committed:
Remove ProcessSyncDataType::HasFullscreenElement
https://bugs.webkit.org/show_bug.cgi?id=287541
rdar://144667685

Reviewed by Tim Nguyen.

We don't need all site isolation processes to know that the top document has
a fullscreen element.  We just need to check if the fullscreen element in this
process is in the top document when a request comes to exit fullscreen.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::populateDocumentSyncDataForNewlyConstructedDocument):
* Source/WebCore/dom/FullscreenManager.cpp:
(WebCore::clearFullscreenFlags):
(WebCore::FullscreenManager::willEnterFullscreen):
(WebCore::FullscreenManager::updatePageFullscreenStatusIfTopDocument): Deleted.
* Source/WebCore/dom/FullscreenManager.h:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::updateProcessSyncData):
(WebCore::Page::setTopDocumentHasFullscreenElement): Deleted.
(WebCore::Page::topDocumentHasFullscreenElement): Deleted.
* Source/WebCore/page/Page.h:
* Source/WebCore/page/ProcessSyncData.in:
* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::requestExitFullScreen):

Canonical link: https://commits.webkit.org/290288@main
